File: backend/app/utils/pdf_generator.py
"""
PDF report generator using ReportLab.
Creates a professional Q&A session report with schema, answers, confidence, and sources.
"""
import io
import re
import base64
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib.colors import HexColor
from reportlab.lib.units import inch
from reportlab.lib.enums import TA_CENTER
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle, Image as RLImage


def get_image_from_base64(base64_str: str, max_width: float = 400, max_height: float = 300):
    try:
        # Strip data URL prefix if present
        if "," in base64_str:
            base64_str = base64_str.split(",", 1)[1]
            
        img_data = base64.b64decode(base64_str)
        img_buffer = io.BytesIO(img_data)
        
        img = RLImage(img_buffer)
        
        # Scale down if necessary maintaining aspect ratio
        img_width = img.imageWidth
        img_height = img.imageHeight
        
        if img_width > max_width or img_height > max_height:
            ratio = min(max_width / float(img_width), max_height / float(img_height))
            img.drawWidth = img_width * ratio
            img.drawHeight = img_height * ratio
            
        return img
    except Exception as e:
        print(f"Error decoding image: {e}")
        return None


def sanitize_text(text: str) -> str:
    """Remove non-ASCII characters and problematic symbols for ReportLab."""
    if not text:
        return ""
    # Remove non-printable characters
    text = re.sub(r"[^\x20-\x7E\n\t]", "", str(text))
    # Escape HTML special chars for ReportLab paragraph rendering
    text = text.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
    
    # Convert basic markdown for ReportLab
    text = re.sub(r'\*\*(.*?)\*\*', r'<b>\1</b>', text) # bold
    text = re.sub(r'\*(.*?)\*', r'<i>\1</i>', text) # italic
    
    # Replace newlines with breaks
    text = text.replace("\n", "<br/>")

    # Truncate very long text
    if len(text) > 4000:
        text = text[:4000] + "... [See full answer in app]"
    return text


def generate_pdf_report(
    messages: list[dict],
    filename: str,
    schema: list[dict],
    semantic_layer=None,
    template_info: dict | None = None,
    attachments: list[dict] | None = None,
) -> bytes:
    """Generate a full PDF report and return as bytes."""
    buffer = io.BytesIO()
    doc = SimpleDocTemplate(
        buffer,
        pagesize=A4,
        leftMargin=0.75 * inch,
        rightMargin=0.75 * inch,
        topMargin=0.75 * inch,
        bottomMargin=0.75 * inch,
    )

    styles = getSampleStyleSheet()

    # Custom styles
    title_style = ParagraphStyle(
        "CustomTitle",
        parent=styles["Title"],
        fontSize=24,
        textColor=HexColor("#3b82f6"),
        spaceAfter=12,
    )
    subtitle_style = ParagraphStyle(
        "Subtitle",
        parent=styles["Normal"],
        fontSize=11,
        textColor=HexColor("#94a3b8"),
        spaceAfter=20,
    )
    question_style = ParagraphStyle(
        "Question",
        parent=styles["Normal"],
        fontSize=11,
        textColor=HexColor("#3b82f6"),
        fontName="Helvetica-Bold",
        spaceBefore=16,
        spaceAfter=4,
    )
    answer_style = ParagraphStyle(
        "Answer",
        parent=styles["Normal"],
        fontSize=10,
        textColor=HexColor("#1f2937"),
        spaceBefore=4,
        spaceAfter=8,
        leftIndent=12,
    )
    code_style = ParagraphStyle(
        "Code",
        parent=styles["Normal"],
        fontSize=8,
        fontName="Courier",
        textColor=HexColor("#1f2937"),
        backColor=HexColor("#f3f4f6"),
        borderPadding=6,
        spaceBefore=6,
        spaceAfter=6,
        leftIndent=12,
        rightIndent=12,
    )
    answer_style = ParagraphStyle(
        "Answer",
        parent=styles["Normal"],
        fontSize=10,
        textColor=HexColor("#1f2937"),
        spaceBefore=4,
        spaceAfter=8,
        leftIndent=12,
    )
    meta_style = ParagraphStyle(
        "Meta",
        parent=styles["Normal"],
        fontSize=8,
        textColor=HexColor("#6b7280"),
        spaceBefore=2,
        spaceAfter=12,
        leftIndent=12,
    )

    story = []

    # ── Title ──────────────────────────────────────────────────────────────
    story.append(Paragraph("DataTalk Analysis Report", title_style))
    story.append(Paragraph(f"Dataset: {sanitize_text(filename)}", subtitle_style))

    # ── Template Info ──────────────────────────────────────────────────────
    if template_info:
        if template_info.get("author"):
            story.append(Paragraph(f"<b>Author:</b> {sanitize_text(template_info['author'])}", styles["Normal"]))
        if template_info.get("company"):
            story.append(Paragraph(f"<b>Company:</b> {sanitize_text(template_info['company'])}", styles["Normal"]))
        if template_info.get("date"):
            story.append(Paragraph(f"<b>Date:</b> {sanitize_text(template_info['date'])}", styles["Normal"]))
        if template_info.get("executive_summary"):
            story.append(Spacer(1, 10))
            story.append(Paragraph("<b>Executive Summary</b>", styles["Heading3"]))
            story.append(Paragraph(sanitize_text(template_info["executive_summary"]), styles["Normal"]))
        story.append(Spacer(1, 20))

    # ── Schema table ────────────────────────────────────────────────────────
    if schema:
        story.append(Paragraph("Dataset Schema", styles["Heading2"]))
        table_data = [["Column", "Type", "Missing %"]]
        for col in schema[:20]:  # Limit to 20 columns
            table_data.append([
                sanitize_text(col["name"]),
                col["type"],
                f"{col['missing_pct']}%",
            ])

        t = Table(table_data, colWidths=[200, 100, 80])
        t.setStyle(TableStyle([
            ("BACKGROUND", (0, 0), (-1, 0), HexColor("#3b82f6")),
            ("TEXTCOLOR", (0, 0), (-1, 0), HexColor("#ffffff")),
            ("FONTSIZE", (0, 0), (-1, -1), 9),
            ("GRID", (0, 0), (-1, -1), 0.5, HexColor("#e5e7eb")),
            ("ROWBACKGROUNDS", (0, 1), (-1, -1), [HexColor("#f9fafb"), HexColor("#ffffff")]),
        ]))
        story.append(t)
        story.append(Spacer(1, 20))

    # ── Q&A pairs ───────────────────────────────────────────────────────────
    if messages:
        story.append(Paragraph("Question &amp; Answer Session", styles["Heading2"]))

        for idx, msg in enumerate(messages):
            role = msg.get("role", "")
            content = sanitize_text(msg.get("content", ""))

            if role == "user":
                story.append(Paragraph(f"Q: {content}", question_style))
            elif role == "assistant":
                story.append(Paragraph(f"A: {content}", answer_style))

                if "image" in msg and msg["image"]:
                    img = get_image_from_base64(msg["image"])
                    if img:
                        story.append(Spacer(1, 10))
                        story.append(img)
                        story.append(Spacer(1, 10))
                # Support matplotlib_image passed from frontend message history
                elif "matplotlib_image" in msg and msg["matplotlib_image"]:
                    img = get_image_from_base64(msg["matplotlib_image"])
                    if img:
                        story.append(Spacer(1, 10))
                        story.append(img)
                        story.append(Spacer(1, 10))

                if attachments:
                    for att in attachments:
                        if att.get("message_index") == idx and att.get("data"):
                            img = get_image_from_base64(att["data"])
                            if img:
                                story.append(Spacer(1, 10))
                                story.append(img)
                                story.append(Spacer(1, 10))

                # Confidence info
                conf = msg.get("confidence")
                if conf:
                    story.append(Paragraph(
                        f"Confidence: {conf.get('score', 'N/A')}% ({conf.get('level', 'N/A')})",
                        meta_style,
                    ))

                # Sources
                sources = msg.get("sources", [])
                if sources:
                    parsed_sources = []
                    for s in sources:
                        if isinstance(s, dict):
                            title = s.get("title") or s.get("name") or s.get("value") or s.get("url") or "Web Source"
                            url = s.get("url")
                            
                            safe_title = sanitize_text(title)
                            if url:
                                safe_url = sanitize_text(url)
                                parsed_sources.append(f'<a href="{safe_url}" color="#3b82f6"><u>{safe_title}</u></a>')
                            else:
                                parsed_sources.append(safe_title)
                        elif isinstance(s, str):
                            parsed_sources.append(sanitize_text(s))
                            
                    source_text = ", ".join(parsed_sources)
                    story.append(Paragraph(
                        f"Sources: {source_text}", meta_style
                    ))

                # SQL query
                sql = msg.get("sql_query")
                if sql:
                    story.append(Paragraph(f"SQL: {sanitize_text(sql)}", meta_style))

    # ── Footer ──────────────────────────────────────────────────────────────
    story.append(Spacer(1, 30))
    story.append(Paragraph(
        "Generated by DataTalk | NatWest Code for Purpose Hackathon",
        ParagraphStyle(
            "Footer",
            parent=styles["Normal"],
            fontSize=8,
            textColor=HexColor("#94a3b8"),
            alignment=TA_CENTER,
        ),
    ))

    doc.build(story)
    return buffer.getvalue()

File: backend-node/src/utils/pdfGenerator.ts
/**
 * PDF report generator using pdfkit.
 */
import PDFDocument from 'pdfkit';

const BLUE = '#3b82f6';
const INDIGO = '#6366f1';
const GREEN = '#10b981';
const AMBER = '#f59e0b';
const GRAY500 = '#6b7280';
const GRAY700 = '#374151';
const GRAY900 = '#111827';
const WHITE = '#ffffff';
const GRAY100 = '#f3f4f6';

function sanitizeText(text: string): string {
  if (!text) return '';
  return String(text)
    .replace(/[^\x20-\x7E\n\t]/g, '')
    .replace(/\*\*(.*?)\*\*/g, '$1')
    .replace(/\*(.*?)\*/g, '$1');
}

function markdownToText(text: string): string {
  return sanitizeText(text)
    .replace(/^#{1,6}\s+/gm, '')
    .replace(/\*\*(.*?)\*\*/g, '$1')
    .replace(/\*(.*?)\*/g, '$1')
    .replace(/`(.*?)`/g, '$1')
    .replace(/^- /gm, '• ');
}

function drawHorizontalRule(doc: PDFKit.PDFDocument, y: number): void {
  doc.moveTo(54, y).lineTo(doc.page.width - 54, y).strokeColor('#e5e7eb').lineWidth(0.5).stroke();
}

function drawTableRow(
  doc: PDFKit.PDFDocument,
  cells: string[],
  widths: number[],
  x: number,
  y: number,
  bgColor: string,
  textColor: string,
  fontSize = 8,
): number {
  const ROW_HEIGHT = 18;
  doc.rect(x, y, widths.reduce((a, b) => a + b, 0), ROW_HEIGHT).fillColor(bgColor).fill();
  doc.fillColor(textColor).fontSize(fontSize);
  let cx = x;
  cells.forEach((cell, i) => {
    doc.text(sanitizeText(cell ?? '').slice(0, 60), cx + 3, y + 4, { width: widths[i] - 6, lineBreak: false });
    cx += widths[i];
  });
  return y + ROW_HEIGHT;
}

function drawTable(
  doc: PDFKit.PDFDocument,
  headers: string[],
  rows: string[][],
  widths: number[],
  startX: number,
  startY: number,
): number {
  let y = drawTableRow(doc, headers, widths, startX, startY, BLUE, WHITE, 8);
  rows.forEach((row, i) => {
    const bg = i % 2 === 0 ? '#f9fafb' : WHITE;
    y = drawTableRow(doc, row, widths, startX, y, bg, GRAY700, 8);
    if (y > doc.page.height - 80) {
      doc.addPage();
      y = 54;
    }
  });
  return y;
}

export function generatePdfReport(opts: {
  messages: Record<string, any>[];
  tables: Record<string, any>;
  template_info?: Record<string, any> | null;
  attachments?: Array<{ message_index?: number; data: string; content_type?: string }> | null;
}): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const { messages, tables, template_info, attachments } = opts;

    const doc = new PDFDocument({ size: 'A4', margin: 54, autoFirstPage: true });
    const chunks: Buffer[] = [];
    doc.on('data', (c: Buffer) => chunks.push(c));
    doc.on('end', () => resolve(Buffer.concat(chunks)));
    doc.on('error', reject);

    const pageWidth = doc.page.width;
    const contentWidth = pageWidth - 108;
    const leftMargin = 54;

    // ─── COVER ───────────────────────────────────────────────────────────────
    doc.moveDown(4);
    doc.fontSize(26).fillColor(BLUE).text('DataTalk', { align: 'center' });
    doc.fontSize(12).fillColor(GRAY500).text('Analysis Report', { align: 'center' });
    doc.moveDown(0.5);

    const tableNames = Object.keys(tables);
    if (tableNames.length > 0) {
      doc.fontSize(10).fillColor(GRAY500).text(`Datasets: ${tableNames.join(' & ')}`, { align: 'center' });
    }

    doc.fontSize(9).fillColor(GRAY500).text(
      `Generated: ${new Date().toLocaleDateString('en-GB', { day: '2-digit', month: 'long', year: 'numeric' })}`,
      { align: 'center' },
    );

    if (template_info) {
      doc.moveDown(1);
      if (template_info.author) doc.fontSize(10).fillColor(GRAY700).text(`Author: ${sanitizeText(template_info.author)}`);
      if (template_info.company) doc.fontSize(10).fillColor(GRAY700).text(`Company: ${sanitizeText(template_info.company)}`);
      if (template_info.executive_summary) {
        doc.moveDown(0.5).fontSize(11).fillColor(GRAY900).text('Executive Summary', { underline: true });
        doc.fontSize(9).fillColor(GRAY700).text(sanitizeText(template_info.executive_summary), { width: contentWidth });
      }
    }

    doc.addPage();

    // ─── DATASET DESCRIPTIONS ────────────────────────────────────────────────
    if (tableNames.length > 0) {
      doc.fontSize(14).fillColor(GRAY900).text('Dataset Descriptions', leftMargin, 54, { underline: false });
      drawHorizontalRule(doc, doc.y + 2);
      doc.moveDown(0.5);

      for (const [tableName, meta] of Object.entries(tables)) {
        const schema: any[] = meta.schema ?? [];
        const quality: any = meta.data_quality ?? meta.dataQuality ?? {};
        const filename = meta.filename ?? tableName;
        const rowCount = meta.row_count ?? meta.rowCount ?? (meta.rows ? meta.rows.length : '—');

        doc.fontSize(11).fillColor(GRAY900).text(`Table: ${tableName}`);
        doc.fontSize(8).fillColor(GRAY700).text(
          `File: ${filename}  |  Rows: ${typeof rowCount === 'number' ? rowCount.toLocaleString() : rowCount}  |  Columns: ${schema.length}  |  Quality Score: ${quality.overall_score ?? 100}%`,
        );
        doc.moveDown(0.3);

        if (schema.length > 0) {
          const schemaRows = schema.slice(0, 30).map((col: any) => [
            col.name ?? '',
            col.type ?? '',
            `${col.missing_pct ?? 0}%`,
            (col.sample_values ?? []).slice(0, 3).join(', ').slice(0, 50),
          ]);
          drawTable(doc, ['Column Name', 'Type', 'Missing %', 'Sample Values'], schemaRows, [130, 70, 65, 165], leftMargin, doc.y);
        }
        doc.moveDown(1);
      }

      doc.addPage();
    }

    // ─── Q&A SESSION ─────────────────────────────────────────────────────────
    const qaMsgs = messages.filter((m) => m.role === 'user' || m.role === 'assistant');

    if (qaMsgs.length > 0) {
      doc.fontSize(14).fillColor(GRAY900).text('Question & Answer Session', leftMargin, 54);
      drawHorizontalRule(doc, doc.y + 2);
      doc.moveDown(0.5);

      let qCount = 0;
      messages.forEach((msg, idx) => {
        if (doc.y > doc.page.height - 100) doc.addPage();

        if (msg.role === 'user') {
          qCount++;
          doc.fontSize(10).fillColor(BLUE).text(`Q${qCount}: ${sanitizeText(msg.content ?? '')}`, { width: contentWidth });
          doc.moveDown(0.3);
        } else if (msg.role === 'assistant') {
          const content = markdownToText(msg.content ?? msg.answer ?? '');
          if (content) {
            doc.fontSize(9).fillColor(GRAY700).text(`A: ${content.slice(0, 2000)}`, { width: contentWidth, lineGap: 2 });
          }

          // Images
          const images: string[] = msg.matplotlib_images ?? (msg.matplotlib_image ? [msg.matplotlib_image] : []);
          for (const b64 of images) {
            try {
              const imgBuf = Buffer.from(b64.includes(',') ? b64.split(',')[1] : b64, 'base64');
              const img = (doc as any).openImage(imgBuf);
              const targetWidth = Math.min(400, contentWidth);
              const scaledHeight = img.height * (targetWidth / img.width);
              
              doc.moveDown(0.5);
              if (doc.y + scaledHeight > doc.page.height - 54) doc.addPage();
              
              const xPos = leftMargin + (contentWidth - targetWidth) / 2;
              doc.image(imgBuf, xPos, doc.y, { width: targetWidth });
              doc.y += scaledHeight;
              doc.moveDown(0.5);
            } catch { /* skip bad image */ }
          }

          // SQL
          if (msg.sql_query) {
            doc.fontSize(7).fillColor(GRAY700).font('Courier')
              .text(`SQL: ${sanitizeText(msg.sql_query).slice(0, 500)}`, { width: contentWidth });
            doc.font('Helvetica');
          }



          // Attachments
          if (attachments) {
            for (const att of attachments) {
              if (att.message_index === idx && att.data) {
                try {
                  const imgBuf = Buffer.from(att.data.includes(',') ? att.data.split(',')[1] : att.data, 'base64');
                  const img = (doc as any).openImage(imgBuf);
                  const targetWidth = Math.min(400, contentWidth);
                  const scaledHeight = img.height * (targetWidth / img.width);
                  
                  doc.moveDown(0.5);
                  if (doc.y + scaledHeight > doc.page.height - 54) doc.addPage();
                  
                  const xPos = leftMargin + (contentWidth - targetWidth) / 2;
                  doc.image(imgBuf, xPos, doc.y, { width: targetWidth });
                  doc.y += scaledHeight;
                  doc.moveDown(0.5);
                } catch { /* skip */ }
              }
            }
          }

          doc.moveDown(0.5);
          drawHorizontalRule(doc, doc.y);
          doc.moveDown(0.5);
        }
      });
    }

    // ─── FOOTER ──────────────────────────────────────────────────────────────
    doc.moveDown(2);
    doc.fontSize(8).fillColor(GRAY500)
      .text('Generated by DataTalk', { align: 'center' });

    doc.end();
  });
}
